### PR TITLE
Use gofmt, not goimports

### PR DESCRIPTION
**Pull request by justinsb**
goimports is slower, should not be necessary (we don't want to add
other imports), and the order of import declarations is no longer
consistent, because goimport groups them alphabetically.

goimports also means that the behaviour of kubebuilder depends on what
is on the GOPATH.

However, goimports will also respect existing import blocks, so we
remain compatible with goimports.

### CODE REVIEW

**Comment by justinsb** [comment]
Although the timing isn't the real reason to merge this (real reason is that goimports walks gopath)...

Before:
```
> time go test -v ./cmd/... ./pkg/... --test.count=1
real    1m57.619s
user    20m54.870s
sys     2m0.850s
```

After:
```
real    0m1.621s
user    0m9.431s
sys     0m2.471s
```

**Comment by DirectXMan12** [comment]
so, there's a problem with this -- we scaffold out imports that may or may not be needed in certain cases.  We'll have to overhaul the scaffolding system, since we currently rely on goimports to do this work for us.

**Comment by droot** [comment]
I am surprised to that timing difference. As Solly pointed out, we sort of exploit `goimport` to sanitize the imports in scaffolded files. Travis failures are related to that.

**Comment by justinsb** [comment]
I guess my view is that if we're using goimports to remove imports, we should prefer to just not add them (some debate there about that increasing complexity of generation, we'll see...).  If we're adding imports, we _really_ shouldn't be doing that, because who knows what `goimports` is going to find on my gopath - we should just add them to the imports block.

But ... this does depend on my being able to get rid of the test failures, which are currently a mixture of the two above.

**Comment by justinsb** [comment]
Travis is happy now!

**Comment by DirectXMan12** [comment]
(FWIW, I'm happy to get rid of goimports if we get it working -- it's been a pain for a while)

**Comment by camilamacedo86** [comment]
Hi @justinsb, 

For you be able to move forward with shows that is missing:
- rebase the branch with the master branch
- then update the testdata by running `make generate-setup` and then `make generate-testdata` to do it by using the Makefile targets.

**Comment by Adirio** [comment]
This code seems to be using go1.12 yet. Please, rebase your changes to master so that changes related to the upgrade to 1.13 are applied.

File: testdata/project-v2/go.mod [hunk 1]
```diff
@@ -1,11 +1,12 @@
 module sigs.k8s.io/kubebuilder/testdata/project-v2
 
-go 1.13
+go 1.12
```
Comment (camilamacedo86) [inline testdata/project-v2/go.mod]: Note that the master branch is using go 1.13. 
In order to collab with the project, you need to use the latest master branch and have go 1.13 locally

File: pkg/scaffold/v2/webhook/webhook.go [hunk 2]
```diff
@@ -91,10 +91,14 @@ var (
 package {{ .Resource.Version }}
 
 import (
+{{ if or .Validating .Defaulting }}
```
Comment (camilamacedo86) [inline pkg/scaffold/v2/webhook/webhook.go]: why the ifs are required to be added here?

File: pkg/scaffold/v1/controller/controller.go [hunk 3]
```diff
@@ -103,14 +103,12 @@ package {{ lower .Resource.Kind }}
...
 import (
 {{ if .Resource.CreateExampleReconcileBody }}	"context"
-	"log"
 	"reflect"
 
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (camilamacedo86) [inline pkg/scaffold/v1/controller/controller.go]: why the schema was removed?

File: pkg/scaffold/v2/webhook/webhook.go (hunk 2 above)
Comment (Adirio) [reply pkg/scaffold/v2/webhook/webhook.go]: Those packages are probably only being used in case you use those flags. As before this change `goimport` tool was being used, this tool removed them, but as this PR swaps `goimport` for `go fmt` imports are not being checked any more and it would scaffold a file with two import compile errors.

**Comment by Adirio** [comment]
IMHO, this PR should not be merged.

You mention some `goimport` drawbacks:
- Being slower: as you said the slowdown is related to `GOPATH`. However, `GOPATH` is only being used for `v1` which is already being deprecated. After that the whole project, scaffolded test and doc packages will be using gomod, which will remove the `GOPATH` dependency and therefore should remove the slowdown.
- Depends on `GOPATH`: not sure if you were still talking about the slowdown issue, but as explained in the previous one, `GOPATH` dependency should not be a thing in a couple months.
- Import order not inconsistency: as you already said yourself, it understands import groups, so once the groups are properly set up though the scaffold (which is a WIP) this disadvantage dissapears.

You also mention some advantages:
- Sanitizes the scaffolding: instead of having to make the scaffolder aware of which imports were already added, it handles duplicates itself.

As all the drawbacks mentioned will be outdated in 2 months time, and the advantage will still hold, I think that swaping `goimport` for `go fmt` would be a step backwards.

**Comment by justinsb** [comment]
I think there's a really important one: correctness.  When we don't specify our imports and rely on goimports, we have no guarantee of what package is going to be imported.

File: pkg/scaffold/v2/webhook/webhook.go (hunk 2 above)
Comment (justinsb) [reply pkg/scaffold/v2/webhook/webhook.go]: Exactly - using gofmt instead of goimports requires us to be precise about what we want to import (good) and not import (less good)

File: pkg/scaffold/v1/controller/controller.go (hunk 3 above)
Comment (justinsb) [reply pkg/scaffold/v1/controller/controller.go]: It doesn't appear to be used - am I missing a spot where it is used?

File: testdata/project-v2/go.mod (hunk 1 above)
Comment (justinsb) [reply testdata/project-v2/go.mod]: Thanks - updated!

File: pkg/scaffold/v1/manager/cmd.go [hunk 4]
```diff
@@ -46,14 +46,15 @@ import (
 	"flag"
 	"os"
 
+	"{{ .Repo }}/pkg/apis"
+	"{{ .Repo }}/pkg/controller"
+	"{{ .Repo }}/pkg/webhook"
+
```
Comment (Adirio) [inline pkg/scaffold/v1/manager/cmd.go]: I would leave this as a separate group but at the end. Reason: these imports are not those of kubebuilder, but of the package we are scaffolding, and thus `{{ .Repo }} ...` imports are the local package ones while the rest are external imports.

Comment (Adirio) [inline pkg/scaffold/v1/manager/cmd.go]: I would leave this as a separate group but at the end. Reason: these imports are not those of kubebuilder, but of the package we are scaffolding, and thus `{{ .Repo }} ...` imports are the local package ones while the rest are external imports.

File: pkg/scaffold/v1/webhook/admissionbuilder.go [hunk 5]
```diff
@@ -81,6 +81,7 @@ package {{ .Type }}
 
 import (
 	{{ .Resource.Group}}{{ .Resource.Version }} "{{ .ResourcePackage }}/{{ .Resource.Group}}/{{ .Resource.Version }}"
+
```
Comment (Adirio) [inline pkg/scaffold/v1/webhook/admissionbuilder.go]: Same

File: pkg/scaffold/v1/webhook/admissionhandler.go [hunk 6]
```diff
@@ -83,7 +83,7 @@ import (
 	"net/http"
 
 	{{ .Resource.Group}}{{ .Resource.Version }} "{{ .ResourcePackage }}/{{ .Resource.Group}}/{{ .Resource.Version }}"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+
```
Comment (Adirio) [inline pkg/scaffold/v1/webhook/admissionhandler.go]: Same

File: testdata/gopath/src/project/pkg/controller/firstmate/firstmate_controller.go [hunk 7]
```diff
@@ -20,14 +20,13 @@ import (
...
 
-	crewv1 "project/pkg/apis/crew/v1"
-
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
+	crewv1 "project/pkg/apis/crew/v1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/firstmate/firstmate_controller.go]: Should be in a new group at the end of import block, same reason as above

File: testdata/gopath/src/project/pkg/controller/firstmate/firstmate_controller_suite_test.go [hunk 8]
```diff
@@ -23,11 +23,10 @@ import (
 	"sync"
 	"testing"
 
-	"project/pkg/apis"
-
 	"github.com/onsi/gomega"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"project/pkg/apis"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/firstmate/firstmate_controller_suite_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/firstmate/firstmate_controller_test.go [hunk 9]
```diff
@@ -20,14 +20,13 @@ import (
...
 
-	crewv1 "project/pkg/apis/crew/v1"
-
 	"github.com/onsi/gomega"
 	"golang.org/x/net/context"
 	appsv1 "k8s.io/api/apps/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	crewv1 "project/pkg/apis/crew/v1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/firstmate/firstmate_controller_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/frigate/frigate_controller.go [hunk 10]
```diff
@@ -19,11 +19,10 @@ package frigate
 import (
 	"context"
 
-	shipv1beta1 "project/pkg/apis/ship/v1beta1"
-
 	appsv1 "k8s.io/api/apps/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	shipv1beta1 "project/pkg/apis/ship/v1beta1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/frigate/frigate_controller.go]: Same

File: testdata/gopath/src/project/pkg/controller/frigate/frigate_controller_suite_test.go [hunk 11]
```diff
@@ -23,11 +23,10 @@ import (
 	"sync"
 	"testing"
 
-	"project/pkg/apis"
-
 	"github.com/onsi/gomega"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"project/pkg/apis"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/frigate/frigate_controller_suite_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/frigate/frigate_controller_test.go [hunk 12]
```diff
@@ -20,13 +20,12 @@ import (
...
 	"time"
 
-	shipv1beta1 "project/pkg/apis/ship/v1beta1"
-
 	"github.com/onsi/gomega"
 	"golang.org/x/net/context"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	shipv1beta1 "project/pkg/apis/ship/v1beta1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/frigate/frigate_controller_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/healthcheckpolicy/healthcheckpolicy_controller.go [hunk 13]
```diff
@@ -19,11 +19,10 @@ package healthcheckpolicy
 import (
 	"context"
 
-	policyv1beta1 "project/pkg/apis/policy/v1beta1"
-
 	appsv1 "k8s.io/api/apps/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	policyv1beta1 "project/pkg/apis/policy/v1beta1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/healthcheckpolicy/healthcheckpolicy_controller.go]: Same

File: testdata/gopath/src/project/pkg/controller/healthcheckpolicy/healthcheckpolicy_controller_suite_test.go [hunk 14]
```diff
@@ -23,11 +23,10 @@ import (
 	"sync"
 	"testing"
 
-	"project/pkg/apis"
-
 	"github.com/onsi/gomega"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"project/pkg/apis"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/healthcheckpolicy/healthcheckpolicy_controller_suite_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/healthcheckpolicy/healthcheckpolicy_controller_test.go [hunk 15]
```diff
@@ -20,13 +20,12 @@ import (
...
 	"time"
 
-	policyv1beta1 "project/pkg/apis/policy/v1beta1"
-
 	"github.com/onsi/gomega"
 	"golang.org/x/net/context"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	policyv1beta1 "project/pkg/apis/policy/v1beta1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/healthcheckpolicy/healthcheckpolicy_controller_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/kraken/kraken_controller.go [hunk 16]
```diff
@@ -19,11 +19,10 @@ package kraken
 import (
 	"context"
 
-	creaturesv2alpha1 "project/pkg/apis/creatures/v2alpha1"
-
 	appsv1 "k8s.io/api/apps/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	creaturesv2alpha1 "project/pkg/apis/creatures/v2alpha1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/kraken/kraken_controller.go]: Same

File: testdata/gopath/src/project/pkg/controller/kraken/kraken_controller_suite_test.go [hunk 17]
```diff
@@ -23,11 +23,10 @@ import (
 	"sync"
 	"testing"
 
-	"project/pkg/apis"
-
 	"github.com/onsi/gomega"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"project/pkg/apis"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/kraken/kraken_controller_suite_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/kraken/kraken_controller_test.go [hunk 18]
```diff
@@ -20,13 +20,12 @@ import (
...
 	"time"
 
-	creaturesv2alpha1 "project/pkg/apis/creatures/v2alpha1"
-
 	"github.com/onsi/gomega"
 	"golang.org/x/net/context"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	creaturesv2alpha1 "project/pkg/apis/creatures/v2alpha1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/kraken/kraken_controller_test.go]: Same

File: testdata/gopath/src/project/pkg/controller/namespace/namespace_controller_suite_test.go [hunk 19]
```diff
@@ -23,11 +23,10 @@ import (
 	"sync"
 	"testing"
 
-	"project/pkg/apis"
-
 	"github.com/onsi/gomega"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"project/pkg/apis"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/controller/namespace/namespace_controller_suite_test.go]: Same

File: testdata/gopath/src/project/pkg/webhook/default_server/namespace/mutating/namespace_update_handler.go [hunk 20]
```diff
@@ -21,6 +21,7 @@ import (
 	"net/http"
 
 	corev1 "k8s.io/api/core/v1"
+
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/webhook/default_server/namespace/mutating/namespace_update_handler.go]: At the end

File: testdata/gopath/src/project/pkg/webhook/default_server/namespace/mutating/update_webhook.go [hunk 21]
```diff
@@ -17,8 +17,9 @@ limitations under the License.
 package mutating
 
 import (
-	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
 	corev1 "k8s.io/api/core/v1"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
```
Comment (Adirio) [inline testdata/gopath/src/project/pkg/webhook/default_server/namespace/mutating/update_webhook.go]: Why do you split these into groups?

**Comment by Adirio** [comment]
> I think there's a really important one: correctness. When we don't specify our imports and rely on goimports, we have no guarantee of what package is going to be imported.

We are not relying on goimport to magically choose the package, we are just relying on it to correctly dedup imports and order them inside each group.

**Comment by justinsb** [comment]
I haven't found a way to tell goimports not to add imports, and because of that we have had omissions sneak into the code e.g. https://github.com/kubernetes-sigs/kubebuilder/pull/951/files#diff-c3eb8f2983d72d8620c4a319e158ef1fR130

If we could figure out how to tell goimports to just remove & reformat - that would be great - I'd probably use that mode in my editor as well!

File: testdata/gopath/src/project/pkg/webhook/default_server/namespace/mutating/update_webhook.go (hunk 21 above)
Comment (justinsb) [reply testdata/gopath/src/project/pkg/webhook/default_server/namespace/mutating/update_webhook.go]: I _think_ it's because corev1 is our target api in this case - interesting corner case!

**Comment by justinsb** [comment]
Looks like tests are flaking, but looks like we're still discussing so not going to worry about it right now

**Comment by Adirio** [comment]
> If we could figure out how to tell goimports to just remove & reformat - that would be great - I'd probably use that mode in my editor as well!

I found a way.

Passing a `nil` options pointer is equivalent to:
```
&imports.Options{
	Comments:  true,
	TabIndent: true,
	TabWidth:  8,
}
```

So if we change it for:
```
&imports.Options{
	AllErrors:  true,
	Comments:   true,
	TabIndent:  true,
	TabWidth:   8,
	FormatOnly: true,
}
```

We would neither add missing packages nor remove unused ones. We would just order and dedup each group separately.

See https://play.golang.org/p/p-mibpxju1d
The `time` package is not added and `github.com/another` is not removed. You can quickly verify that both changes would be made if you swap `FormatOnly` to `false`.

It still fails to dedup' packages found in two import groups, but that also happens with that option set to false so I guess it is an improvement.

**Comment by Adirio** [comment]
@justinsb Are you ok if I include my proposed alternative in #1187? It will require to add some of the changes you have made (removing non-needed and adding missing imports)

**Comment by Adirio** [comment]
@mengqiy This PR is outdated due to #1187. IMHO it can be closed.

**Comment by DirectXMan12** [comment]
closing as per https://github.com/kubernetes-sigs/kubebuilder/pull/951#issuecomment-572911081
